File: golang-devildyw/middle/L19删除链表的倒数第N个结点.go
package main

func main() {
	var head = &ListNode{
		Val:  1,
		Next: nil,
	}
	removeNthFromEnd(head, 1)
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	// 虚拟节点搞起来
	dummyHead := &ListNode{
		Val:  -1,
		Next: head,
	}
	cur := dummyHead
	pre := dummyHead

	for i := 0; i < n; i++ {
		cur = cur.Next
	}
	// 一定是 cur.Next !=nil 这样遍历下去 pre就会停在需要删除节点的前一个位置 这样刚好方便删除
	for cur.Next != nil {
		cur = cur.Next
		pre = pre.Next
	}

	pre.Next = pre.Next.Next

	return dummyHead.Next
}

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L19删除链表的倒数第N个结点.java
package top.dyw.leetcode.Hot100;


public class L19删除链表的倒数第N个结点 {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummyNode = new ListNode(-1,head);
        ListNode pre = dummyNode, cur = dummyNode.next;
        while(n > 0) {
            cur = cur.next;
            n--;
        }


        while (cur!=null) {
            pre = pre.next;
            cur = cur.next;
        }

        pre.next = pre.next.next;

        return dummyNode.next;

    }
}
